best_practices/README.rst

```diff
@@ -0,0 +1,142 @@
+
+This document uses terminology (MUST, SHOULD, etc) from `RFC 2119`_.
+
+.. _RFC 2119: https://www.ietf.org/rfc/rfc2119.txt
+
+----------
+Background
+----------
+
+Incompatibilities between packages published on the `Python Package Index (PyPI)`_
+has been a long standing issue. Diamond dependencies are a common problem where
+a package or application has dependencies that are, themselves, dependant on
+incompatible versions of shared packages.
+has been a long standing issue. Diamond dependency (aka dependency hell) is a
+common problem that a package’s dependencies depend on different and
+incompatible versions of the shared packages.
+
+.. _Python Package Index (PyPI): https://pypi.org/
+
+Incompatibilities between packages can occur when:
+
+- A package makes breaking API changes and doesn't follow `Semantic Versioning`_
+- A package has a pinned dependency version which conflicts with other dependencies.
+- A package depends on outdated dependencies.
+- A package is depends on deprecated dependencies.
+
+.. _Semantic Versioning: https://semver.org/
+
+This guide is a list of best practices that Python package authors can follow
+to help reduce future incompatibilities. Google-sponsored projects are expected
+to follow these guidelines but other projects may benefit from them as well.
+
+-----
+Tools
+-----
+
+    To detect and prevent version incompatibilities between Google OSS Python
+    packages, we provide a set of tooling to ensure we are compatible with
+    ourselves. And the tooling can also run checks for any Google owned package that wants to
+    know whether it is self-compatible or compatible with a list of other packages.
+    Below lists our tooling:
+
+Dependency Compatibility Dashboard
+----------------------------------
+
+The `dashboard`_ shows the compatibility status of a list of Google owned OSS
+Python packages.
+
+.. _dashboard: https://googlecloudplatform.github.io/cloud-opensource-python/
+
+Badge Server
+------------
+
+The badge server runs checks and generates an svg format badge image to show the
+status of a given package. Supported usage includes:
+
+- Self compatibility
+- Google-wise compatibility
+- Dependency version status
+- One badge for all kinds of checks above
+
+For more details please refer to `here`_.
+
+.. _here: https://github.com/GoogleCloudPlatform/cloud-opensource-python/tree/master/badge_server
+
+----------------------
+Python Packaging Rules
+----------------------
+
+Package Versioning
+------------------
+
+We should use semantic versioning for all Google OSS Python distribution
+packages. Semantic versioning requires that given a version number
+`MAJOR.MINOR.PATCH`, increment the:
+
+    * MAJOR version when you make incompatible API changes.
+    * MINOR version when you add functionality in a backwards-compatible manner.
+    * PATCH version when you make backwards-compatible bug fixes.
+
+Requirements:
+
+- GA(Generally Available) libraries must conform to semantic versioning.
```

Comment:
Make "Generally Available" a link to https://cloud.google.com/terms/launch-stages